src/lib/cfa/sudoers.rb

```diff
@@ -0,0 +1,237 @@
+    #
+    # @see CFA::BaseModel#initialize
+    # @see #fix_collection_names
+    # @see #set_start_placer
+    def load
+      super
+
+      fix_collection_names(data)
+      @start_placer = set_start_placer
+    end
+
+    # Return all users specifications
+    #
+    # @return [Array<Hash>]
+    def users_specs
+      current_users.reduce([]) do |specs, user|
+        spec = {}
+        values = user[:value]
+
+        spec["user"] = values["user"]
+        spec["host"] = values["host_group"]["host"]
+        spec["commands"] = values["host_group"].collection("command").map do |command|
+          command.is_a?(AugeasTreeValue) ? command.value : command
+        end
+        spec["run_as"] = values["host_group"]["command[]"]["runas_user"]
+        spec["tag"] = values["host_group"]["command[]"]["tag"]
+
+        specs << spec.reject { |_, v| v.nil? }
+      end
+    end
+
+    # Remove all users specifications
+    def clean_users
+      data.delete(USERS_MATCHER)
+    end
+
+    # Add a new basic user specification
+    #
+    #
+    # A user specification determines which commands a user may run (and as what user) on specified
+    # hosts (who may run what, for short).
+    #
+    # Its basic structure is “who where = (as_whom) what”.
+    #
+    # E.g., with
+    #
+    #   root            ALL = (ALL) ALL
+    #   %wheel          ALL = (ALL) ALL
+    #
+    #   We let root and any user in group wheel run any command on any host as any user.
+    #
+    # See the sudoers manpage for more details.
+    #
+    # @param user_spec [Hash] a user specification
+    # @option user_spec [String] "user" who can run the command
+    # @option user_spec [String] "host" where the user can run the command
+    # @option user_spec [Array<String>] "commands" commands that the user can run
+    # @option user_spec [String] "run_as" as whom the commands will be run
+    # @option user_spec [String] "tag" the tag associated to the command (e.g. NOPASSWD)
+    def add_user(user_spec)
+      user = AugeasTree.new
+      host_group = AugeasTree.new
+
+      host_group.add("host", user_spec["host"])
+
+      commands = user_spec["commands"].reduce([]) do |commands, command|
+        augeas_command = AugeasTree.new
+        host_group.add("command[]", AugeasTreeValue.new(augeas_command, command))
+        commands << augeas_command
+      end
+
+      # For the time being, "runas_user" and "tag" options are added only to the first command,
+      # which means that they apply to all commands.
+      run_as = user_spec["run_as"].to_s.strip
+      tag = user_spec["tag"].to_s.strip
+      commands[0].add("runas_user", run_as) unless run_as.empty?
+      commands[0].add("tag", tag) unless tag.empty?
+
+      user.add("user", user_spec["user"])
+      user.add("host_group", host_group)
+
+      data.add("spec[]", user, placer)
+    end
+
+    private
+
+    attr_accessor :start_placer
+
+    # Return the current users collection
+    #
+    # @return [Array<Hash>]
```

Comment:
for sure define that it is just copy, so do not do modification on top of it. Another think is that I do not believe that it is really Hashes in that array. It looks more like AugeasTree. So I expect it will be modified to that hash or am I wrong?

Reply:
It looks as Array of Hashes

```
(byebug) data.select(USERS_MATCHER).class
Array
(byebug) data.select(USERS_MATCHER).first.class
Hash
(byebug) data.select(USERS_MATCHER).first
{:key=>"spec[]", :value=>#<CFA::AugeasTree:0x000055cf7d43aab0 @data=[{:key=>"user", :value=>"ALL", :orig_key=>"user", :operation=>:keep}, {:key=>"host_group", :value=>#<CFA::AugeasTree:0x000055cf7d43a6a0 @data=[{:key=>"host", :value=>"ALL", :orig_key=>"host", :operation=>:keep}, {:key=>"command[]", :value=>#<CFA::AugeasTreeValue:0x000055cf7d43a100 @tree=#<CFA::AugeasTree:0x000055cf7d43a2e0 @data=[{:key=>"runas_user", :value=>"ALL", :orig_key=>"runas_user", :operation=>:keep}]>, @value="ALL", @modified=false>, :orig_key=>"command", :operation=>:keep}]>, :orig_key=>"host_group", :operation=>:keep}, {:key=>"#comment", :value=>"WARNING! Only use this together with 'Defaults targetpw'!", :orig_key=>"#comment", :operation=>:keep}]>, :orig_key=>"spec[1]", :operation=>:keep}
```

Reply:
yep, but it is internal augeas tree hash, which I think is not good idea to expose. Especially see parts like orig_key or operation. also that :key and :value is probably not what you want to expose to users of your model.

Reply:
> yep, but it is internal augeas tree hash, which I think is not good idea to expose

It is an internal/private method, used by `#users_specs` which only exposes a Ruby Hash

Reply:
ok, then at least please document it in method documentation(that it is augeas tree internal structure that can change) and also mention to not modify that structures.